File: pyNA/src/noise_src_py/epnl.py
import pdb
import openmdao
import openmdao.api as om
import numpy as np
from tqdm import tqdm
from pyNA.src.settings import Settings

def epnl(self, t_o: np.ndarray, pnlt: np.ndarray, C: np.ndarray = np.ones(1)) -> np.float64:
    """
    Compute effective perceived noise level.

    :param t_o: Observer time [s]
    :type t_o: np.ndarray [n_t']
    :param pnlt: perceived noise level, tone corrected [PNdB]
    :type pnlt: np.ndarray [n_t']
    :param C: pnlt tone correction [dB]
    :type C: np.ndarray [n_t'] settings.N_f']

    :return: epnl [EPNdB]
    :rtype: Float64
    """
    
    # Load options
    settings = self.options['settings']

    # Interpolate time, pnlt and C
    n_ip = np.int64(np.ceil((t_o[-1]-t_o[0])/0.5))

    t_ip = np.zeros(n_ip)
    for i in np.arange(n_ip):
        t_ip[i] = t_o[0] + i*0.5

    pnlt_ip = np.interp(t_ip, t_o, pnlt)

    # Compute max. PNLT
    pnltm = np.max(pnlt_ip)

    # Check tone band-sharing
    i_max = np.where(pnlt_ip == pnltm)[0][0]
    if settings.bandshare:

        C_ip = np.zeros((n_ip, settings.N_f))
        for j in np.arange(settings.N_f):
            C_ip[:,j] = np.interp(t_ip, t_o, C[:,j])

        if i_max == 0 or i_max == 1:
            i_left_bandshare = 0
        else:
            i_left_bandshare = i_max-2

        if i_max == np.shape(C_ip)[0]-1 or i_max == np.shape(C_ip)[0]-2:
            i_right_bandshare = np.shape(C_ip)[0]
        else:
            i_right_bandshare = i_max + 3

        C_bandshare = np.zeros(i_right_bandshare-i_left_bandshare)
        for i, k in enumerate(np.arange(i_left_bandshare, i_right_bandshare)):
            C_bandshare[i] = np.max(C_ip[k, :])

        if C_bandshare[2] < np.mean(C_bandshare):
            pnltm = pnltm - C_bandshare[2] + np.mean(C_bandshare)

    # Compute max. PNLT point (k_m)
    I = np.where(pnlt_ip > pnltm - 10.)

    # ICAO Annex 16 procedures (p132 - App. 2-18)
    f_int = 10. ** (pnlt_ip / 10.)

    # Compute integration bounds
    if pnltm > 10:
        i_1 = I[0][0]
        if np.abs(pnlt_ip[i_1] - (pnltm - 10)) > np.abs(pnlt_ip[i_1 - 1] - (pnltm - 10)):
            i_1 = i_1 - 1

        i_2 = I[0][-1]
        if i_2 < pnlt_ip.shape[0] - 1:
            if np.abs(pnlt_ip[i_2] - (pnltm - 10)) > np.abs(pnlt_ip[i_2 + 1] - (pnltm - 10)):
                i_2 = i_2 + 1

        D = 10 * np.log10(np.sum(f_int[i_1:i_2])) - pnltm - 10 * np.log10(20)

    else:
        D = 10 * np.log10(np.sum(f_int)) - pnltm - 10 * np.log10(20)

    # Compute EPNL
    epnl = pnltm + D

    return epnl

File: pyNA/src/noise_src_jl/epnl.jl
using PCHIPInterpolation

function f_epnl(x::Union{Array, ReverseDiff.TrackedArray}, settings)

    # x = [t_o, level]
    # y = epnl
    
    # Number of time steps
    n_t = Int64(size(x)[1]/2)

    # Interpolate time, pnlt and C
    n_ip = Int64(ceil((x[n_t]-x[1])/settings["epnl_dt"]))
    
    t_ip = zeros(eltype(x), (n_ip,))
    for i in range(1, n_ip, step=1)
        t_ip[i] = x[1] + (i-1)*settings["epnl_dt"]
    end

    # Interpolate the data
    f_ipnlt_interp = PCHIPInterpolation.Interpolator(x[1:n_t], x[n_t+1:end])
    pnlt_ip = f_ipnlt_interp.(t_ip)

    # Compute max. PNLT
    pnltm = maximum(pnlt_ip)

    # Check tone band-sharing
    # I_max = findall(x->x==pnltm, pnlt_ip)
    # if settings["epnl_bandshare"]
    #     continue
    # end
    
    # Compute max. PNLT point (k_m)
    I = findall(x->x>pnltm - 10.,pnlt_ip)

    # ICAO Annex 16 procedures (p132 - App. 2-18)
    f_int = 10 .^(pnlt_ip / 10.)

    # Compute duration correction
    if pnltm > 10
        # Compute integration bounds: lower bound
        i_1 = I[1]
        if i_1 != 1 
            if abs(pnlt_ip[i_1] - (pnltm - 10)) > abs(pnlt_ip[i_1 - 1] - (pnltm - 10))
                i_1 = i_1 - 1
            end
        end

        # Compute integration bounds: upper bound
        i_2 = I[end]
        if i_2 != size(pnlt_ip)[1]
            if (abs(pnlt_ip[i_2] - (pnltm - 10))) > (abs(pnlt_ip[i_2 + 1] - (pnltm - 10)))
                i_2 = i_2 + 1
            end
        end
         
        # Duration correction
        D = 10 * log10(sum(f_int[i_1:i_2])) - pnltm - 10 * log10(10. / settings["epnl_dt"])
    else
        # Duration correction
        D = 10 * log10(sum(f_int)) - pnltm - 10 * log10(10. / settings["epnl_dt"])
    end
    
    # Compute EPNL
    epnl = pnltm + D

    return epnl

end

function f_epnl!(epnl::Array, x::Union{Array, ReverseDiff.TrackedArray}, settings)

    # x = [t_o, level]
    # y = epnl
    
    # Number of time steps
    n_t = Int64(size(x)[1]/2)

    # Interpolate time, pnlt and C
    n_ip = Int64(ceil((x[n_t]-x[1])/settings["epnl_dt"]))
    
    t_ip = zeros(eltype(x), (n_ip,))
    for i in range(1, n_ip, step=1)
        t_ip[i] = x[1] + (i-1)*settings["epnl_dt"]
    end

    # Interpolate the data
    f_ipnlt_interp = PCHIPInterpolation.Interpolator(x[1:n_t], x[n_t+1:end])
    pnlt_ip = f_ipnlt_interp.(t_ip)

    # Compute max. PNLT
    pnltm = maximum(pnlt_ip)

    # Check tone band-sharing
    # I_max = findall(x->x==pnltm, pnlt_ip)
    # if settings["epnl_bandshare"]
    #     continue
    # end
    
    # Compute max. PNLT point (k_m)
    I = findall(x->x>pnltm - 10.,pnlt_ip)

    # ICAO Annex 16 procedures (p132 - App. 2-18)
    f_int = 10 .^(pnlt_ip / 10.)

    # Compute duration correction
    if pnltm > 10
        # Compute integration bounds: lower bound
        i_1 = I[1]
        if i_1 != 1 
            if abs(pnlt_ip[i_1] - (pnltm - 10)) > abs(pnlt_ip[i_1 - 1] - (pnltm - 10))
                i_1 = i_1 - 1
            end
        end

        # Compute integration bounds: upper bound
        i_2 = I[end]
        if i_2 != size(pnlt_ip)[1]
            if (abs(pnlt_ip[i_2] - (pnltm - 10))) > (abs(pnlt_ip[i_2 + 1] - (pnltm - 10)))
                i_2 = i_2 + 1
            end
        end
         
        # Duration correction
        D = 10 * log10(sum(f_int[i_1:i_2])) - pnltm - 10 * log10(10. / settings["epnl_dt"])
    else
        # Duration correction
        D = 10 * log10(sum(f_int)) - pnltm - 10 * log10(10. / settings["epnl_dt"])
    end
    
    # Compute EPNL
    epnl .= pnltm + D

end

f_epnl_fwd! = (y,x)->f_epnl!(y, x, settings)